### PR TITLE
Add `CBORGenerator.Feature.LENIENT_UTF_ENCODING` for lenient handling of Unicode surrogate pairs on writing

### PR DESCRIPTION
**Pull request by guillaumebort**
If enabled, the generator will output the Unicode Replacement Character for invalid unicode sequence (invalid surrogate chars in the Java String) instead of failing with an IllegalArgumentException.

Also this PR remove the code duplication between `_shortUTF8Encode2` and `_encode2`.

### CODE REVIEW

File: cbor/src/main/java/com/fasterxml/jackson/dataformat/cbor/CBORGenerator.java [hunk 1]
```diff
@@ -61,7 +66,14 @@
...
          * written at the beginning of a new document.
          */
-        WRITE_TYPE_HEADER(false)
+        WRITE_TYPE_HEADER(false),
+
+        /**
+         * Feature that determines if an invalid surrogate encoding found in the
+         * incoming String should fail with an exception or silently be outputed
+         * as the Unicode 'REPLACEMENT CHARACTER' (U+FFFD)
+         */
```
Comment (cowtowncoder) [inline cbor/src/main/java/com/fasterxml/jackson/dataformat/cbor/CBORGenerator.java:75]: Could you add `@since` annotation (but see discussion on version to include in)

File: cbor/src/main/java/com/fasterxml/jackson/dataformat/cbor/CBORGenerator.java [hunk 2]
```diff
@@ -138,6 +150,8 @@ public int getMask() {
 
     protected boolean _cfgMinimalInts;
 
+    protected boolean _cfgLenientUnicodeEncoding;
```
Comment (cowtowncoder) [inline cbor/src/main/java/com/fasterxml/jackson/dataformat/cbor/CBORGenerator.java:153]: Brief javadoc would be nice (I know, some other entries do not have it yet but... )

File: cbor/src/main/java/com/fasterxml/jackson/dataformat/cbor/CBORGenerator.java [hunk 3]
```diff
@@ -1452,20 +1425,43 @@ private final int _encode2(int i, int outputPtr, String str, int len,
             }
             // Yup, a surrogate pair
             if (c > SURR1_LAST) { // must be from first range; second won't do
-                _throwIllegalSurrogate(c);
+                if (_cfgLenientUnicodeEncoding) {
```
Comment (cowtowncoder) [inline cbor/src/main/java/com/fasterxml/jackson/dataformat/cbor/CBORGenerator.java:1428]: One small suggestion: change `_throwIllegalSurrogate` to contain this check, and either throw exception or return surrogate value. That way rarely needed logic contained in that method, bit more compact on call site.

**Comment by cowtowncoder** [comment]
Ok, first of all, thank you for the contribution! I am bit concerned about the fact that some code is trying to output invalid Unicode content, essentially, but as long as that is documented and user has to explicitly enable said feature I think that is fine.

I added some small notes in PR itself, but 2 bigger questions:

1. Since "master" is for 3.0, which might be far out, you may want to instead make PR against 2.12. I can handle merging it forward to master (wrt API changes)
2. I'll need to have a look but my main concern is with performance: since this is an edge case, it should have no measurable effect on good case (i.e. content does not have invalid surrogate characters). I can probably test it myself but thought I'll mention it -- I think changes do affect main loops.

File: cbor/src/main/java/com/fasterxml/jackson/dataformat/cbor/CBORGenerator.java [hunk 4]
```diff
@@ -1356,81 +1377,33 @@ private final int _encode(int outputPtr, char[] str, int i, int end) {
...
         } while (++i < end);
         return outputPtr - outputStart;
     }
 
-    /**
-     * Helper method called when the whole character sequence is known to fit in
-     * the output buffer, but not all characters are single-byte (ASCII)
-     * characters.
-     */
-    private final int _shortUTF8Encode2(char[] str, int i, int end,
```
Comment (cowtowncoder) [inline cbor/src/main/java/com/fasterxml/jackson/dataformat/cbor/CBORGenerator.java:1371]: Curious as to why this was removed? Or did it just get moved and diff is confused.

File: cbor/src/main/java/com/fasterxml/jackson/dataformat/cbor/CBORGenerator.java [hunk 5]
```diff
@@ -1356,81 +1377,33 @@ private final int _encode(int outputPtr, char[] str, int i, int end) {
...
-
     private final int _encode(int outputPtr, String str, int len) {
         final byte[] outBuf = _outputBuffer;
         final int outputStart = outputPtr;
 
         for (int i = 0; i < len; ++i) {
             int c = str.charAt(i);
             if (c > 0x7F) {
-                return _encode2(i, outputPtr, str, len, outputStart);
+                return _encode2(i, outputPtr, str.toCharArray(), len, outputStart);
```
Comment (cowtowncoder) [inline cbor/src/main/java/com/fasterxml/jackson/dataformat/cbor/CBORGenerator.java:1394]: So this change was made because.... ?

File: cbor/src/main/java/com/fasterxml/jackson/dataformat/cbor/CBORGenerator.java (hunk 4 above)
Comment (guillaumebort) [reply cbor/src/main/java/com/fasterxml/jackson/dataformat/cbor/CBORGenerator.java:1371]: It has been removed because this code was duplicated: `_shortUTF8Encode2` and `_encode2` were basically the same with the difference that one was taking a `String` as an argument and the other one was taking a `char[]`. Since this code contains some really non trivial logic I thought it would be better to not duplicate it.

File: cbor/src/main/java/com/fasterxml/jackson/dataformat/cbor/CBORGenerator.java (hunk 5 above)
Comment (guillaumebort) [reply cbor/src/main/java/com/fasterxml/jackson/dataformat/cbor/CBORGenerator.java:1394]: I have changed the signature of `_encode2` to take a char array instead so I could replace usage of `_shortUTF8Encode2` by `_encode2` and kill `_shortUTF8Encode2`.

**Comment by guillaumebort** [comment]
> 1. Since "master" is for 3.0, which might be far out, you may want to instead make PR against 2.12. I can handle merging it forward to master (wrt API changes)

Ok fair enough I will base the PR on 2.12

File: cbor/src/main/java/com/fasterxml/jackson/dataformat/cbor/CBORGenerator.java (hunk 1 above)
Comment (guillaumebort) [reply cbor/src/main/java/com/fasterxml/jackson/dataformat/cbor/CBORGenerator.java:77]: Done

File: cbor/src/main/java/com/fasterxml/jackson/dataformat/cbor/CBORGenerator.java (hunk 2 above)
Comment (guillaumebort) [reply cbor/src/main/java/com/fasterxml/jackson/dataformat/cbor/CBORGenerator.java:160]: Done

File: cbor/src/main/java/com/fasterxml/jackson/dataformat/cbor/CBORGenerator.java (hunk 3 above)
Comment (guillaumebort) [reply cbor/src/main/java/com/fasterxml/jackson/dataformat/cbor/CBORGenerator.java]: Yes good idea, done.

**Comment by cowtowncoder** [comment]
@guillaumebort Apologies for slow follow-up here: now back to getting this merge for 2.12.
One thing I'd need, if I hadn't yet asked would be CLA. It's here:

https://github.com/FasterXML/jackson/blob/master/contributor-agreement.pdf

and usually easiest to print, fill & sign, scan/photo, email to `info` at fasterxml dot com.
Only needs to be done once before the first contribution. Apologies if we already got one and I somehow missed it.

**Comment by guillaumebort** [comment]
> One thing I'd need, if I hadn't yet asked would be CLA. It's here:

Thanks! I need to handle that with the legal team at my company and I do it ASAP.

**Comment by cowtowncoder** [comment]
Sounds good -- we have both individual CLA that I linked earlier (and used by most contributors), as well as Corporate CLA (CCLA), at

https://github.com/FasterXML/jackson/blob/master/contributor-agreement-corporate.txt

if that makes more sense.

**Comment by guillaumebort** [comment]
Should be ok now, my company (Datadog) sent a a signed copy of the Corporate CLA over.

**Comment by cowtowncoder** [comment]
@guillaumebort For some reason I don't see one yet? I assume it'd be sent to `info@fasterxml.com`?

**Comment by cowtowncoder** [comment]
Received the CLA.

File: cbor/src/main/java/com/fasterxml/jackson/dataformat/cbor/CBORGenerator.java (hunk 5 above)
Comment (cowtowncoder) [reply cbor/src/main/java/com/fasterxml/jackson/dataformat/cbor/CBORGenerator.java:1469]: Ok, so this would add another allocation (str.toCharArray()). I think I'll undo this change & manually merge other changes.

**Comment by cowtowncoder** [comment]
Ended up merging this manually, hence closing PR, feature is in, tests, will be included in `2.12.0-rc2`.
Thank you again for contributing this! Might make sense to add similar support for Smile, and perhaps other backends too.
